File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const res = [];
   
   for (let i = 0; i < nums.length; i++) {

       if (i === 0) {
           res.push(1);
           continue;
       }

       res.push(res[i - 1] * nums[i - 1])

   }

   let post

   for (let i = nums.length - 1; i >= 0; i--) {
       if (i === nums.length - 1) {
           post = nums[i]
           continue
       }

       res[i] *= post
       post *= nums[i]
   }

    return res
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        ans = [1] * length
        for i in range(1, length):
            ans[i] *= ans[i - 1] * nums[i - 1]
        postfix = 1
        for i in range(length - 1, -1, -1):
            if i == length - 1:
                postfix *= nums[i]
            else:
                ans[i] *= postfix
                postfix *= nums[i]
        return ans